### PR TITLE
Add PSA crypto driver delegation for AEAD single shot PSA APIs - 1

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
Add PSA crypto driver delegation for AEAD single shot PSA APIs following the design defined in psa-crypto-implementation-structure.md as well as basic testing of the dispatch to driver/software fallback.

In terms of testing of the driver interface and its integration into the PSA application interface implementation, with this PR:

- all AEAD encrypt/decrypt cryptographic operations are triggered by a call to the AEAD encrypt/decrypt driver interface. The call is the same for hardware and software implementations. The AEAD driver interface is thus exercised by all the PSA unit tests that test AEAD encryption/decryption.

- the all.sh test_psa_crypto_drivers component runs all PSA unit tests with all AEAD encryptions/decryptions handled by the transparent test driver.

This PR does not add support for AEAD operations in the the all.sh test_psa_crypto_config_basic component. This will be done when #4246 is merged as additional commits in this PR or in a separate PR.

## Status
READY

## Requires Backporting
NO, PSA only.

## Steps to test or reproduce
CI only.

### CODE REVIEW

File: library/psa_crypto_aead.c
Comment (stevew817) [inline library/psa_crypto_aead.c:35]: Might as well clean this up while at it. cipher_info is only used in the setup call, and therefore doesn't need to be in the operation context.

File: library/psa_crypto_aead.c
Comment (stevew817) [inline library/psa_crypto_aead.c:49]: Might as well clean this up while at it. I don't believe it is necessary to keep track of `core_alg` in the context, since all call trees contain the algorithm.

File: library/psa_crypto_aead.c
Comment (stevew817) [inline library/psa_crypto_aead.c:50]: Might as well clean this up while at it. `full_tag_length` is only used inside the setup call, and therefore doesn't need to be in the operation context.

File: library/psa_crypto_aead.c
Comment (stevew817) [inline library/psa_crypto_aead.c:51]: Might as well clean this up while at it. `tag_length` is encoded in the algorithm, so it doesn't need to be stored separately.

File: library/psa_crypto_aead.c
Comment (stevew817) [inline library/psa_crypto_aead.c:70]: Might as well clean this up while at it, and add `mbedtls_chachapoly_free` to the switchcase (which arguably should've been there already). The only reason it doesn't fail CI is because the current chachapoly context doesn't allocate memory, so the memory leakage checkers don't understand that the free function isn't being called.

File: tests/suites/test_suite_psa_crypto_driver_wrappers.data
Comment (stevew817) [inline tests/suites/test_suite_psa_crypto_driver_wrappers.data:242]: Why is this test case annotated with `INSUFFICIENT_MEMORY`, but expects to return `PSA_SUCCESS`?

File: library/psa_crypto.c
Comment (ronald-cron-arm) [reply library/psa_crypto.c:3914]: I cherry-picked your commit "Do validation on the algorithm argument in AEAD" but changes the returned error code to PSA_ERROR_NOT_SUPPORTED as there is this in the PSA spec:
"PSA_ERROR_NOT_SUPPORTED alg is not supported or is not an AEAD algorithm."

File: library/psa_crypto_aead.c
Comment (ronald-cron-arm) [reply library/psa_crypto_aead.c:49]: But we will probably need it for multi-part operation.

File: library/psa_crypto_aead.c
Comment (ronald-cron-arm) [reply library/psa_crypto_aead.c:51]: Ditto, we will probably need it for multi-part operations.

File: library/psa_crypto_aead.c
Comment (ronald-cron-arm) [reply library/psa_crypto_aead.c:70]: Indeed, done.

File: tests/suites/test_suite_psa_crypto_driver_wrappers.data
Comment (ronald-cron-arm) [reply tests/suites/test_suite_psa_crypto_driver_wrappers.data:242]: The return code was wrong.

**Comment by ronald-cron-arm** [comment]
@stevew817 I believe I addressed all your comments and the result can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-aead-1).
Then I rebased because of the conflicts with the merge of #4246.
The resolution of the conflicts were just about replacing MBEDTLS_*_C symbols by the MBEDTLS_PSA_BUILTIN_* corresponding ones.

**Comment by stevew817** [comment]
Now that #4246 in in, care to add AEAD to the all.sh psa_crypto_config and psa_crypto_driver tests?

**Comment by ronald-cron-arm** [comment]
My plan was to do that when rebasing/reworking #4240. See [comment](https://github.com/ARMmbed/mbedtls/pull/4240#issuecomment-806539963).

**Review by tom-daubney-arm — CHANGES_REQUESTED** [review]
This PR looks mostly fine to me, however one small thing I noticed in library/psa_crypto_aead.c

The order of the conditional compilation macros is not consistent between CCM and GCM. In the `struct` declaration (Ln 33 onward), `#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)` comes before `#if defined(MBEDTLS_PSA_BUILTIN_ALG_GCM)`. 
The same is true in the function `psa_abort_internal` (Ln 54 onward) and the function `psa_aead_setup`.

However, in the functions `mbedtls_psa_aead_encrypt` (Ln 165 onward) and `mbedtls_psa_aead_decrypt` (Ln 275 onward), the order swaps and we have `#if defined(MBEDTLS_PSA_BUILTIN_ALG_GCM)` coming before `#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)`.

It would be good if this ordering was consistent and I don't think there is a good reason for them to be the way they are right now. I suggest swapping the order of the conditional compilation statements in the `mbedtls_psa_aead_encrypt` (Ln 165 onward) and `mbedtls_psa_aead_decrypt` (Ln 275 onward) functions such that the `#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)` statements come first.

**Comment by ronald-cron-arm** [comment]
I addressed @tom-daubney-arm comment with an additional commit and the result can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-aead-2). Then I rebased onto development and resolved the conflicts: only minor conflicts due to the merge of the cipher driver delegation rework #4256.

**Comment by ronald-cron-arm** [comment]
After the rebase from commit "psa: aead: Remove key slot from operation context" the static and inline function `psa_get_key_slot_bits()` in psa_crypto.c was not used anymore and CI jobs failing. I've updated the commit aforementioned with the removal of `psa_get_key_slot_bits()`.

File: library/psa_crypto_aead.c
Comment (stevew817) [reply library/psa_crypto_aead.c:48]: I'm not going to block the PR over this, but for the record I'll say that I think that if multi-part isn't going to be implemented soon than it would still make sense to remove the unused elements until the implementation is being done. Having structs with unreferenced/unused items is a bit icky.

**Review by stevew817 — APPROVED** [review]
This is good to go in IMO.

File: library/psa_crypto_aead.c
Comment (gilles-peskine-arm) [reply library/psa_crypto_aead.c:48]: [Multipart AEAD is just around the corner](https://github.com/ARMmbed/mbedtls/pull/4338).

Comment (stevew817) [reply library/psa_crypto_aead.c:48]: Thanks, I hadn't seen that yet. I guess it deserves an explicit link to this PR?

Comment (gilles-peskine-arm) [reply library/psa_crypto_aead.c:48]: Given that this PR looks almost ready but multipart AEAD is still in a draft state, I expect that @paul-elliott-arm will rebase multipart AEAD once driver delegation is merged.

**Review by tom-daubney-arm — APPROVED** [review]
Thanks for the changes. LGTM.

**Pull request merged**
